src/apify/_charging.py

```diff
@@ -0,0 +1,286 @@
+                    'Using the ACTOR_TEST_PAY_PER_EVENT environment variable is only supported '
+                    'in a local development environment'
+                )
+
+            self._pricing_model = 'PAY_PER_EVENT'
+
+        self._client = client
+        self._charging_log_dataset: Dataset | None = None
+
+        self._charging_state: dict[str, ChargingStateItem] | None = None
+        self._pricing_info: dict[str, PricingInfoItem] = {}
+
+        self._not_ppe_warning_printed = False
+
+    async def init(self) -> None:
+        """Initialize the charging manager - this is called by the `Actor` class and shouldn't be invoked manually."""
+        self._charging_state = {}
+
+        if self._is_at_home:
+            # Running on the Apify platform - fetch pricing info for the current run.
+
+            if self._actor_run_id is None:
+                raise RuntimeError('Actor run ID not found even though the Actor is running on Apify')
+
+            run = run_validator.validate_python(await self._client.run(self._actor_run_id).get())
+            if run is None:
+                raise RuntimeError('Actor run not found')
+
+            if run.pricing_info is not None:
+                self._pricing_model = run.pricing_info.pricing_model
+
+                if run.pricing_info.pricing_model == 'PAY_PER_EVENT':
+                    for event_name, event_pricing in run.pricing_info.pricing_per_event.actor_charge_events.items():
+                        self._pricing_info[event_name] = PricingInfoItem(
+                            price=event_pricing.event_price_usd,
+                            title=event_pricing.event_title,
+                        )
+
+                    self._max_total_charge_usd = run.options.max_total_charge_usd or self._max_total_charge_usd
+
+            for event_name, count in (run.charged_event_counts or {}).items():
+                price = self._pricing_info.get(event_name, PricingInfoItem(Decimal(), title='')).price
+                self._charging_state[event_name] = ChargingStateItem(
+                    charge_count=count,
+                    total_charged_amount=count * price,
+                )
+
+        if not self._is_at_home and self._pricing_model == 'PAY_PER_EVENT':
+            # We are not running on the Apify platform, but PPE is enabled for testing - open a dataset that
+            # will contain a log of all charge calls for debugging purposes.
+
+            if self._purge_charging_log_dataset:
+                dataset = await Dataset.open(name=self.LOCAL_CHARGING_LOG_DATASET_NAME)
+                await dataset.drop()
+
+            self._charging_log_dataset = await Dataset.open(name=self.LOCAL_CHARGING_LOG_DATASET_NAME)
+
+    async def charge(self, event_name: str, count: int = 1) -> ChargeResult:
+        """Charge for a specified number of events - sub-operations of the Actor.
+
+        This is relevant only for the pay-per-event pricing model.
+        """
```

Comment:
missing arg docstring, especially if we have 2 events with different meanings now :smile: